### PR TITLE
add and update tests for new behaviour with empty backups

### PR DESCRIPTION
**Pull request by uhoreg**

### CODE REVIEW

**Comment by uhoreg** [comment]
corresponds to https://github.com/matrix-org/synapse/pull/4123

**Comment by uhoreg** [comment]
The tests in the corresponding synapse PR run fine, so I'm not sure why they're failing here, and the test results seem to be empty.  Unless it was because I did a force-push to the synapse branch while these tests were running... :man_shrugging:

**Comment by richvdh** [comment]
> Aborting test run due to failure to start test server

-- for some reason, synapse didn't start.

@matrixbot retest this please

File: tests/41end-to-end-keys/07-backup.pl [hunk 1]
```diff
@@ -40,6 +38,44 @@
...
+      })->then( sub {
+         my ( $content ) = @_;
+
+         log_if_fail "Content", $content;
+
+         assert_deeply_eq( $content, {"rooms" => {}});
+
+         Future->done(1);
+         matrix_get_backup_key( $user, '', '', 'bogusversion');
+      })->main::expect_http_404;
```
Comment (richvdh) [inline tests/41end-to-end-keys/07-backup.pl]: can you use `expect_m_not_found`, to check the error code as well as the http code?

File: tests/41end-to-end-keys/07-backup.pl [hunk 2]
```diff
@@ -40,6 +38,44 @@
...
+   requires => [ $fixture, qw( can_create_backup_version ) ],
+
+   do => sub {
+      my ( $user ) = @_;
+      my $version;
+
+      matrix_get_key_backup_info( $user )->then( sub {
+         my ( $content ) = @_;
+
+         log_if_fail "Content", $content;
```
Comment (richvdh) [inline tests/41end-to-end-keys/07-backup.pl]: could we have a more descriptive description than "Content" ? (likewise below)

File: tests/41end-to-end-keys/07-backup.pl [hunk 3]
```diff
@@ -40,6 +38,44 @@
...
+
+         $version = $content->{version};
+
+         matrix_get_backup_key( $user, '!notaroom', 'notassession', $version);
+      })->main::expect_http_4xx
```
Comment (richvdh) [inline tests/41end-to-end-keys/07-backup.pl]: why is this a 4xx rather than something more specific? Also, please can we have a comment to say what we are testing here?

File: tests/41end-to-end-keys/07-backup.pl [hunk 4]
```diff
@@ -40,6 +38,44 @@
...
+      my $version;
+
+      matrix_get_key_backup_info( $user )->then( sub {
+         my ( $content ) = @_;
+
+         log_if_fail "Content", $content;
+
+         $version = $content->{version};
+
+         matrix_get_backup_key( $user, '!notaroom', 'notassession', $version);
```
Comment (richvdh) [inline tests/41end-to-end-keys/07-backup.pl]: missing a space before the `)`

File: tests/41end-to-end-keys/07-backup.pl [hunk 5]
```diff
@@ -393,9 +435,19 @@ =head2 matrix_get_backup_key
 sub matrix_get_backup_key {
    my ( $user, $room_id, $session_id, $version ) = @_;
 
+   my $uri;
+
+   if ($session_id) {
```
Comment (richvdh) [inline tests/41end-to-end-keys/07-backup.pl]: probably better to check for definedness rather than truthiness (`if ( defined $session_id )`) to distinguish between an absent param and an empty one.

Also, spaces inside parens please. (I don't make the rules...)

File: tests/41end-to-end-keys/07-backup.pl [hunk 6]
```diff
@@ -40,6 +38,44 @@
...
+         my ( $content ) = @_;
+
+         log_if_fail "Content", $content;
+
+         $version = $content->{version};
+
+         matrix_get_backup_key( $user, '!notaroom', 'notassession', $version);
+      })->main::expect_http_4xx
+      ->then( sub {
+         matrix_get_backup_key( $user, '!notaroom', '', $version);
```
Comment (richvdh) [inline tests/41end-to-end-keys/07-backup.pl]: Again, it would be useful to have some more words to explain what we are testing and what we expect the result to be.

File: tests/41end-to-end-keys/07-backup.pl [hunk 7]
```diff
@@ -384,18 +435,28 @@ sub matrix_backup_keys {
 
 =head2 matrix_get_backup_key
 
-   matrix_get_backup_key( $user, $room_id, $session_id, $version )
+   matrix_get_backup_key( $user, $version, $room_id, $session_id )
 
 Send keys to a given key backup version
```
Comment (richvdh) [inline tests/41end-to-end-keys/07-backup.pl]: this looks wrong. could you fix it while you're here?

File: tests/41end-to-end-keys/07-backup.pl [hunk 8]
```diff
@@ -40,6 +38,53 @@
...
+         # than an error when nothing has been backed up yet
+         matrix_get_backup_key( $user, $version );
+      })->then( sub {
+         my ( $content ) = @_;
+
+         log_if_fail "Get all keys: ", $content;
+
+         assert_deeply_eq( $content, { "rooms" => {} } );
+
+         Future->done(1);
```
Comment (richvdh) [inline tests/41end-to-end-keys/07-backup.pl]: we don't need this
